### PR TITLE
Cherry-picks `step` fixes from Pavel's PR

### PR DESCRIPTION
**Pull request by djspiewak**
This is a straight extension of #326, I just wanted to create a separate PR in order to keep the review clean.  I basically just cherry picked some commits from #318 and fixed the merge conflicts.  The result is a) more tests in `WyeSpec`, and b) a working `step` function!

BTW, I'd also like to petition we rename `runAsync` to `stepAsync`.  Its name implies that it runs the process all the way to the end, which is not at all what actually happens.

### CODE REVIEW

**Comment by djspiewak** [comment]
It turns out that `njoin` and `wye` were assuming different invariants from `runAsync`!  This explains a lot, actually…  Anyway, `wye` assumes that `runAsync` can invoke the callback with an empty sequence of elements.  Which is intuitive, since it's a `Seq` and not a `NEL`.  Anyway, there is an explicit test for this case in `wye` and a special fast-path fallthrough if we land in that case, which is what I would expect.  *However,* `njoin` explicitly assumes that `runAsync` will not ever call back with an empty sequence!  This is ultimately what makes the `step` fix tricky.

Anyway, it turns out to be really easy (and remarkably efficient) to resolve this problem on `njoin`'s end, rather than trying to contort `runAsync` to only produce non-empty results.  It just gives rise to an additional actor message inside of `njoin`, but all of the existing state machinery is sufficient to catch the case and either abort the process early or continue on without inhibition (as in fact, one would expect given the other things `njoin` has to juggle).

**Comment by pchlupacek** [comment]
@djspiewak yeah, good catch. I remember we had it somewhere !. Thanks a lot.

**Comment by pchlupacek** [comment]
@djspiewak ah ok, your repo :-)

**Comment by alissapajer** [comment]
@djspiewak just verified that this PR contains all of the wye specs that I had originally written that reproduced various bugs.

**Comment by alissapajer** [comment]
We've fixed a number of other race conditions. At this point there are no known flaws, other than a test that would be nice to add, testing a particular race condition.

So @djspiewak and I are set for this to be merged.

File: src/main/scala/scalaz/stream/Process.scala [hunk 1]
```diff
@@ -1192,93 +1207,261 @@ object Process extends ProcessInstances {
...
+                  interrupted.get() match {
+                    case ptr @ -\/(int) => {
+                      if (interrupted.compareAndSet(ptr, \/-(cause))) {
+                        int()
+                      } else {
+                        referencedInterrupt(cause)
+                      }
+                    }
+
+                    case \/-(_) => ()    // ?????
```
Comment (pchlupacek) [inline src/main/scala/scalaz/stream/Process.scala]: Can you perhaps comment the question marks?

**Comment by pchlupacek** [comment]
@djspiewak @alissapajer I just add that one single comment to be updated. We perhaps need in future to revisit still the runAsync et all implementation, but I would say great work overall and LGTM.

File: src/main/scala/scalaz/stream/Process.scala (hunk 1 above)
Comment (djspiewak) [inline src/main/scala/scalaz/stream/Process.scala]: I'll elaborate on them.  Basically, that case can only occur if we are interrupted multiple times with different causes.  We always drop the later causes on the floor in that case.

**Comment by pchlupacek** [comment]
@djspiewak are we ready to merge? not sure on this travis build error?

**Comment by djspiewak** [comment]
@pchlupacek We're ready to merge.  I filed an issue for the error; it is unrelated to `runAsync`.

**Comment by djspiewak** [comment]
Travis is being buggy again.  I'm pretty sure extending the timeout fixed the legitimate issue (which I believe was caused by travis being slow).  If we're all in consensus that this is good-to-go, I'm going to pull the trigger on the merge.

**Pull request merged**